Remote-Hosting of Certain Package Data

tbowers7 opened the pull request:
The main thrust of this PR is to reduce the size of the PyPI-hosted release packages by omitting various package-included data such as `reid_arxiv` and `sensfunc` files.  

Major Changes:
* Add download/caching for various package-included data using AstroPy's [`download_file()`](https://docs.astropy.org/en/stable/api/astropy.utils.data.download_file.html#astropy.utils.data.download_file) functionality.  Remove data files from package export in the `MANIFEST.in` file.
* Use this framework to directly pull Telluric Atmospheric Grid files from the cloud into the user's cache without separate download from Google Drive.
* Also use this framework to save the output of `pypeit_identify` (`wvarxiv.fits`) into the user's cache as a way to directly and quickly use manual wavelength identifications in reductions.
* Add unit tests to ensure the downloading introduced in this PR continues to work.

Minor Changes:
* Tweak the `pypeit_identify` GUI to display the spectrum on log scale by default (linear scale selectable by command line argument).
* Address #1345, since I was tweaking wavelength stuff anyways.


Further Discussion:
Especially since most users will not need most of the files in the `pypeit.data` directories (and they are somewhat large), this PR adds functionality to download them from the GitHub repo at runtime when needed.

The downloading and caching process takes advantage of AstroPy's [`download_file()`](https://docs.astropy.org/en/stable/api/astropy.utils.data.download_file.html#astropy.utils.data.download_file) functionality, along with user-space caching.  Downloaded files are nominally kept in `~/.pypeit/cache`, but this can be changed using the [Astropy configuration system](https://docs.astropy.org/en/stable/config/index.html).  For files being kept out of the release distribution, PypeIt now checks first the package directory, then the cache, then downloads the file if not already found.

Taking advantage of this new framework, this PR also revamps the `pypeit_install_telluric` script to download a specified Telluric Atmospheric Grid file from the cloud into the cache for use.  This alleviates the need for separately downloading said files from the Google Drive and installing them into the package directory.

A feature requested by `LDT/DeVeny` users is the ability to quickly and easily use manual wavelength calibrations for their reductions.  This PR uses the new caching functionality to save a copy of the output `wvarxiv.fits` from `pypeit_identify` (with date/time in the filename) for direct reuse with the `full_template` method.

Unit Tests Pass:
```
============================= test session starts ==============================
platform linux -- Python 3.9.12, pytest-7.1.1, pluggy-1.0.0
3.7.0, doctestplus-0.12.0, hypothesis-6.41.0
collected 329 items                                                            
======================= 329 passed in 1868.46s (0:31:08) =======================
```

Dev Suite (branch: `develop`) is getting caught on the newer `keck_lris_red_mark4` files -- also a `gemini_gnirs` error "No matching objects for OBJ0659-DET01", but otherwise passes.  May need to run the `dev-suite` in the cloud.

tbowers7 committed:
Adjust the `pypeit_identify` GUI to match QA; default scaling = log

In particular, use a logarithmic plot for the spectrum (as does the QA plot) to
identify weak lines more easily by eye.  Also, add axis labels for good measure.

Retain previous behavior through new `--linear` argument.

	modified:   pypeit/core/gui/identify.py
	modified:   pypeit/scripts/identify.py

tbowers7 committed:
Address Issue #1345

Remove deprecated WavelengthSolutionPar keywords `wv_cen` and `disp`.

	modified:   pypeit/par/pypeitpar.py
	modified:   pypeit/spectrographs/gtc_osiris.py
	modified:   pypeit/spectrographs/wht_isis.py

tbowers7 committed:
First steps toward keeping package data remotely

To reduce the package size on PyPI, keep package data files (such as
`reid_arxiv`) remotely.  This branch will introduce code to check on the local
machine for a requested file, and download it from the remote server if not
present.

As most users only need a small number of `reid_arxiv` files, this keeps things
lean.

This commit simply moves the construction of the full path for the `reid_arxiv`
to `pypeit.data.utils`.  More to come...

	modified:   pypeit/core/wavecal/waveio.py
	modified:   pypeit/data/utils.py
	modified:   pypeit/scripts/show_arxiv.py

tbowers7 committed:
Surprisingly simple solution

This downloading setup places the file into the `~/.pypeit` directory (or other,
if specified differently in the user's astropy configuration; see
https://docs.astropy.org/en/stable/config/index.html#astropy-config).

Next steps will be to verify file names, and set up the github URL to point to
the tag matching the Pypeit `version`.

	modified:   pypeit/data/utils.py

tbowers7 committed:
Download/Cache for reid_arxiv, sensfunc, and skisim files

Create path-returning functions for reid_arxiv, sensfunc, and skisim files
from the `pypeit/data` directory.

If the file is in the package directory, functions return that path.  This
should only be the case for from-source installations (i.e., developers).
For all end-user installs (e.g., from `pip` or `conda`), these files will not
be included in the distribution for compactness of the package.  For these
users, use AstroPy's `download_file` functionality to download and cache these
files (currently from the GitHub repo) for use.  By default, the cache is in
the user's home directory -- but can be configured through AstroPy's interface.

Next steps will include documentation changes to describe how all this works.

	modified:   MANIFEST.in
	modified:   pypeit/core/coadd.py
	modified:   pypeit/core/flux_calib.py
	modified:   pypeit/core/wavecal/waveio.py
	modified:   pypeit/data/utils.py
	modified:   pypeit/sensfilearchive.py
	modified:   pypeit/wavemodel.py

tbowers7 committed:
If working in a `dev` version, copy downloaded files to package directory

	modified:   pypeit/data/utils.py

tbowers7 committed:
Fold in the Telluric Grids

Download / cache the telluric atmospheric grid files from "the cloud" (links
forthcoming from @profxj).

Modify the `pypeit_install_telluric` script to do the downloading from the cloud
directly into the cache (without the previous need to download from Google Drive
then create a symlink in the package directory).

	modified:   pypeit/data/utils.py
	modified:   pypeit/scripts/install_telluric.py
	modified:   pypeit/scripts/show_arxiv.py

tbowers7 committed:
Save manual wavelength identification to cache; Doc changes

Save the output of `pypeit_identify` to the PypeIt cache for direct use.  This
includes small modifications to the `pypeit_identify` script to pass down
appropriate information for the ad hoc filename.

Update the `pypeit_install_telluric` script to download the TelGrid files into
the cache.

Update the "Additional Data" and "Wavelength Calibration" documentation to
reflect the changes in this commit.

	modified:   doc/installing.rst
	modified:   doc/wave_calib.rst
	modified:   pypeit/core/gui/identify.py
	modified:   pypeit/data/utils.py
	modified:   pypeit/par/pypeitpar.py
	modified:   pypeit/scripts/identify.py
	modified:   pypeit/scripts/install_telluric.py

tbowers7 committed:
Allow install of local telluric grid in the cache

One concern from @jhennawi about removing absolute paths to telluric grid files
was that power users who create their own telluric grids wouldn't be able to
just state the full path to their creations.  This commit now allows the
installation in the cache of local files.

	modified:   pypeit/scripts/install_telluric.py

tbowers7 committed:
Remove `use_local` stubs

With the addition of the `pypeit.data.write_file_to_cache()` routine, we can
write local files to the PypeIt cache.  As such, removing the `use_local` stubs
present in the various `get_*_filepath()` routines.

	modified:   pypeit/data/utils.py

tbowers7 committed:
Add S3 URL for telluric grids in a file for easy update/download

	modified:   pypeit/data/utils.py
	new file:   pypeit/data/s3_url.txt
	modified:   pypeit/scripts/install_telluric.py

tbowers7 committed:
Merge branch 'develop' into remote_package_data

tbowers7 committed:
Merge branch 'develop' into remote_package_data

tbowers7 committed:
Small tweaks following testing of this branch

	modified:   pypeit/core/gui/identify.py
	modified:   pypeit/data/utils.py

tbowers7 committed:
Add unit tests for package data routines

In particular, check that downloading non-local package data still works.

	modified:   pypeit/data/utils.py
	new file:   pypeit/tests/test_pkgdata.py

tbowers7 committed:
Add user instructions for using TelGrids not in the cloud

For telluric grids not hosted in the cloud, the code gracefully exits and
instructs the user to download the data from the Google Drive.

	modified:   doc/installing.rst
	modified:   pypeit/data/utils.py

tbowers7 committed:
Additional user information

Specify users call 'method = full_template' when using the manual wavelength
calibration output of `pypeit_identify`.

	modified:   pypeit/core/gui/identify.py

tbowers7 committed:
Merge branch 'develop' into remote_package_data

tbowers7 committed:
Fix unit test for CI compatability

	modified:   pypeit/tests/test_pkgdata.py

tbowers7 commented:
The CI tests are revealing some issues with pathnames and filenames in this new scheme.  I will need to address them before this PR can be reviewed.

tbowers7 committed:
Add ability to copy cache to package directory, if requested

This commit addresses a couple different CI failures in the parent PR.

	modified:   MANIFEST.in
	modified:   pypeit/data/utils.py
	modified:   pypeit/sensfilearchive.py
	modified:   pypeit/tests/test_sensfilearchive.py

tbowers7 committed:
Update gtc_osiris.py

Accidentally had commented out a `lamps` parameter.  Fixed now.

tbowers7 committed:
Merge branch 'develop' into remote_package_data

profxj commented on pypeit/core/gui/identify.py:
```diff
@@ -269,7 +276,14 @@ def initialise(cls, arccen, lamps, slits, slit=0, par=None, wv_calib_all=None,
         plt.subplots_adjust(bottom=0.05, top=0.85, left=0.05, right=0.65)
         ax.add_line(spec)
-        ax.set_ylim((0.0, 1.1 * spec.get_ydata().max()))
+        if y_log:
```
Nice added feature

profxj commented on pypeit/par/pypeitpar.py:
```diff
@@ -2476,18 +2476,6 @@ def __init__(self, reference=None, method=None, echelle=None, ech_fix_format=Non
-        # Backwards compatibility with basic and semi_brute algorithms
-        defaults['wv_cen'] = 0.0
-        dtypes['wv_cen'] = float
-        descr['wv_cen'] = 'Central wavelength. Backwards compatibility with basic and ' \
```
well pruned

profxj commented on pypeit/spectrographs/gtc_osiris.py:
```diff
@@ -276,58 +276,58 @@ def config_specific_par(self, scifile, inp_par=None):
         if self.get_meta_value(scifile, 'dispname') == 'R300B':
-            par['calibrations']['wavelengths']['wv_cen'] = 4405.
-            par['calibrations']['wavelengths']['disp'] = 4.96
+            # par['calibrations']['wavelengths']['wv_cen'] = 4405.
```
@rcooke-ast -- you might confirm this is ok

rcooke-ast replied:
Yes, fine to comment it out, but maybe don't delete just yet - thanks!

profxj commented on pypeit/spectrographs/wht_isis.py:
```diff
@@ -171,8 +171,8 @@ def default_pypeit_par(cls):
         par['calibrations']['wavelengths']['sigdetect'] = 10.0
-        par['calibrations']['wavelengths']['wv_cen'] = 4859.0
-        par['calibrations']['wavelengths']['disp'] = 0.2
+        # par['calibrations']['wavelengths']['wv_cen'] = 4859.0
```
also a quick check for @rcooke-ast

rcooke-ast replied:
Yes, this can be deleted

profxj commented on pypeit/tests/test_pkgdata.py:
```diff
@@ -0,0 +1,55 @@
+"""
```
Thanks for the tests!!

profxj reviewed (approved):
Great stuff.

Should we consider a test in the DevSuite too?
Or do you feel what you've generated is enough?

Am approving either way.  But maybe wait for @rcooke-ast to comment on the elmination of semi-brute and basic.  Thanks!

rcooke-ast commented on pypeit/core/gui/identify.py:
```diff
@@ -269,7 +276,14 @@ def initialise(cls, arccen, lamps, slits, slit=0, par=None, wv_calib_all=None,
+            ax.set_ylim( (max(1., spec.get_ydata().min()),
+                        4.0 * spec.get_ydata().max()))
+        else:
+            ax.set_ylim((0.0, 1.1 * spec.get_ydata().max()))
```
Perhaps be explicit, and add:
`ax.set_yscale('linear')`

I can see reasons why we might want to toggle between log and linear during the fitting (e.g. the user... likely me... forgets to put the command line flag)

tbowers7 replied:
Added.

rcooke-ast commented on pypeit/core/gui/identify.py:
```diff
@@ -701,17 +715,27 @@ def store_solution(self, final_fit, binspec, rmstol=0.15,
+                msgs.info(f"Your arxiv solution has been cached.{msgs.newline()}"
+                          f"Use 'reid_arxiv = {cachename}' and{msgs.newline()}"
+                          f"'method = full_template' in your PypeIt{msgs.newline()}"
+                          "Reduction File to utilize this wavelength solution.")
```
Should we add the explicit lines of code the user needs to copy paste into their reduction file?

```
[calibrations]
  [[wavelengths]]
    reid_arxiv = {cachename}
    method = full_template
```

tbowers7 replied:
Added.

rcooke-ast commented on pypeit/par/pypeitpar.py:
```diff
@@ -2554,7 +2542,7 @@ def valid_methods():
         """
         # TODO: Remove from this list anything that is not valid!
-        return ['simple', 'semi-brute', 'basic', 'holy-grail', 'identify', 'reidentify',
+        return ['simple', 'holy-grail', 'identify', 'reidentify',
```
Yes! Be gone `basic` and `semi-brute`!  I don't think they are even used...

tbowers7 replied:
They are not used... I did a thorough check of all non-deprecated code and found they are commented out.  I chose not to delete the commented code sections entirely, but maybe we can consider moving them to `deprecated/` somehow and remove the commented code from `pypeit/core/arc.py`.

tbowers7 commented:
> Should we consider a test in the DevSuite too?
> Or do you feel what you've generated is enough?

@profxj : I think an independent `dev-suite` run in the cloud will be very useful given the new feature of needing to download certain data from a remote server.  I found several issues just with the CI tests, and I would appreciate having a separate `dev-suite` run not on my own machine.

tbowers7 commented:
@tepickering : I'm interested in your view of how I implemented AstroPy's [`download_file()`](https://docs.astropy.org/en/stable/api/astropy.utils.data.download_file.html#astropy.utils.data.download_file) functionality, and if it conforms to standard usage.

tbowers7 committed:
Addressing @rcooke-ast PR comments

	modified:   pypeit/core/gui/identify.py
	modified:   pypeit/spectrographs/wht_isis.py

Commit:
yscale

tepickering commented on pypeit/data/utils.py:
```diff
@@ -121,6 +132,413 @@ class Paths(metaclass=Paths_meta):
+    """write_file_to_cache Use `astropy.utils.data` to save local file to cache
+
+    This function writes a local file to the PypeIt cache as if it came from a
+    remote server.  This is useful for being able to use lcoally created files
```
small typo

tepickering reviewed (approved):
really great stuff! i heartily approve!

tepickering commented:
> @tepickering : I'm interested in your view of how I implemented AstroPy's [`download_file()`](https://docs.astropy.org/en/stable/api/astropy.utils.data.download_file.html#astropy.utils.data.download_file) functionality, and if it conforms to standard usage.

looks good to me. like i said in the slack convo, ours is the exact use case that it was built for and you're using it exactly as intended.

tbowers7 committed:
Fixing typo; doc updates

	modified:   doc/help/pypeit_identify.rst
	modified:   doc/help/pypeit_install_telluric.rst
	modified:   doc/pypeit_par.rst
	modified:   doc/wave_calib.rst
	modified:   pypeit/data/utils.py

tbowers7 committed:
Merge branch 'develop' into remote_package_data

tbowers7 committed:
Doc tweak

	modified:   doc/pypeit_par.rst
	modified:   pypeit/par/pypeitpar.py

Commit:
yscale

Commit:
typo

tbowers7 committed:
Maintain cache name for TelGrid files even if S3 URL changes

Utilize the `sources` feature of AstroPy's `download_file()` to maintain the
cache name (encoded as the source URL) for Tellurig Grid files while pointing
to updated URLs, should the need arise.  This will prevent users from having
to re-download TelGrid files (4-8GB) just because the S3 server URL changed,
because the cache is tied to the source URL.  The cache name is now enshrined
(carved in stone, as it were) with the current (2022-04-24) S3 URL, but that
is hashed and the user won't be bothered with it.

If changes in the S3 URL are required, simply update the URL in
`pypeit/data/s3_url.txt` on the `release` branch, and the code will point to the
correct file, caching it under the old URL name.

	modified:   pypeit/data/utils.py
	modified:   pypeit/tests/test_pkgdata.py

tbowers7 committed:
Merge branch 'develop' into remote_package_data

tbowers7 committed:
Merge pull request #1386 from pypeit/remote_package_data_yscale

Remote package data yscale

tbowers7 committed:
The S3 "permanent" URL need not be an actual place, just a name

Use a permenent, but fake, URL for the S3 cloud cache index

	modified:   pypeit/data/utils.py

tbowers7 commented:
Cloud run of `dev-suite` passes with only 1 (well-documented) error:
```
And it is our (not so favorite):
End of Log:
  File "/PypeIt/pypeit/pypmsgs.py", line 185, in error
    raise PypeItError(msg)
pypeit.pypmsgs.PypeItError: No matching objects for OBJ0211-DET01.  Odds are you input the wrong OBJID
```

I will add a little more documentation about how to run PypeIt without internet access (_i.e._, how to pre-download necessary files) before merging.

tbowers7 committed:
Add ability to "pre-cache" needed GitHub-hosted package data

At the request of @profxj, this commit adds the ability to pre-cache GitHub-
hosted package data in the event that a user wants to use PypeIt while not
connected to the internet.

Adds a new console script `pypeit_cache_github_data` that takes the name of
a spectrograph and caches the necessary package data from GitHub.  This may be
done rather than downloading/caching on the fly during a PypeIt run.

In order to quickly and easily find the necessary files on GitHub, this commit
adds a new dependency for PypeIt: `pygithub`

	modified:   environment.yml
	modified:   pypeit/data/utils.py
	new file:   pypeit/scripts/cache_github_data.py
	modified:   setup.cfg

tbowers7 commented:
@kbwestfall : I would appreciate your eye on these last few changes, in addition to any comments you have on the PR overall.

Some more doc changes forthcoming.

tbowers7 committed:
Reduce the # of calls to the GitHub API

For unauthenticated users (i.e., our code), the GitHub API has a limit on the
number of requests in a given time span.  The CI environment overload this
quite quickly, so finding ways to reduce the number of calls to the API.

	modified:   pypeit/data/utils.py
	modified:   pypeit/scripts/cache_github_data.py

tbowers7 committed:
Re-working how the GitHub API is used to minimize calls

Now, use it once to produce a base URL, then build the individual file paths
as we did before.

	modified:   pypeit/data/utils.py
	modified:   pypeit/tests/test_pkgdata.py

tbowers7 committed:
Realized we don't need many calls to the GitHub API at all

	modified:   pypeit/data/utils.py
	modified:   pypeit/scripts/cache_github_data.py

kbwestfall commented on doc/installing.rst:
```diff
@@ -229,43 +240,36 @@ Atmospheric Model Grids
+instruct you to download the file separately from the `PypeIt dev-suite Google Drive`_.
+Users may select any of the files in the Google Drive for their telluric correction,
+download them sepatately, then install them using the ``--local`` option to
+``pypeit_install_telluric``.
```
Awesome!

kbwestfall commented on doc/wave_calib.rst:
```diff
@@ -181,20 +181,26 @@ quit the GUI to see if you want to save the solution. Note,
+                  following block in your PypeIt Reduction File:
+                  [calibrations]
+                     [[wavelengths]]
+                        reid_arxiv = manual_ldt_deveny_20220422T1006.fits
```
Should this be `wvarxiv.fits` to match the example in the text?

tbowers7 replied:
I've addressed this in 2 ways.
  1. No longer writing to generic `wvarxiv.fits`, but instead writing to `wvarxiv_{spec}_{datetime}.fits` in the local directory.
  2. I've corrected the text above this code block to indicate what is saved in the cache is labeled with the spectrograph and timestamp of creation for uniqueness.  This is in case a user is doing manual IDs for multiple setups and wants to be able to use all of them.

kbwestfall commented on pypeit/data/utils.py:
```diff
@@ -121,6 +133,430 @@ class Paths(metaclass=Paths_meta):
 
+# Remote-fetch functions for package data not distributed via PyPI ===========#
+def get_reid_arxiv_filepath(arxiv_file, copy_to_pkgdir=False):
+    """get_reid_arxiv_filepath Return the full path to the `reid_arxiv` file
```
You don't need the function name in the docstring.

kbwestfall commented on pypeit/data/utils.py:
```diff
@@ -121,6 +133,430 @@ class Paths(metaclass=Paths_meta):
+          The base filename of the `reid_arxiv` file to be located
+        copy_to_pkgdir: bool, optional
+          Force copy any cached downloaded file into the package directory
+          [Default: False]
```
I don't think the above will render correctly.  Check the syntax and distinction between Google and Numpy styles: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/#google-vs-numpy .  You can pick either, but the syntax has to consistently conform to one of the two within any given docstring.

kbwestfall commented on pypeit/sensfilearchive.py:
```diff
@@ -67,11 +67,12 @@ class DEIMOSSensFileArchive(SensFileArchive):
     spec_name = "keck_deimos"
 
-    def get_archived_sensfile(self, fitsfile):
+    def get_archived_sensfile(self, fitsfile, unit_test=False):
```
Can we change this to be explicitly `copy_to_pkgdir`?  I definitely appreciate the intention to test this with a unit test.  But, as much as possible, functions should not "know" whether or not they're being executed as part of a unit test.

kbwestfall reviewed (changes requested):
Amazing work, @tbowers7 !!  Thanks for all your effort on this.  I have a couple minor requests, though:

 - Please check the docstring syntax.  I don't think Sphinx will correctly render what you have in many of the `data.utils` functions.
 - I realize I'm in the minority, but the use of `as` is a pet peeve of mine.  Unless it's an "industry standard" (like `import numpy as np`), I think we should almost always avoid using this.  Specifically, instead of `from astropy.utils import data as astropy_data`, I'd prefer `import astropy.utils.data`; and instead of `from pypeit import __version__ as pypeit_version`, I'd prefer `from pypeit import __version__`.  And then propagate the relevant changes.

Three other thoughts/questions not necessarily for this PR:

- Similar to your excellent changes to the telluric installation script, is it useful to update the QL masters installation script?
- Can you remind me of the use case for the `copy_to_pkgdir` option?  Is it possible/useful to instead have a symlink installed in the package directory?  I'm wary of having multiple files on disk, particularly for the large telluric files.
- For the telluric files, can we eventually get all the telluric files in the cloud?  I think I remember that being the goal, but I don't remember the hang-up.

profxj commented:
Regarding @kbwestfall 's comments on the QL Masters' files, 
  I am going to generate Cloud scripts for `release` and `develop`
 that will run `build_ql_masters` and push the outputs to `s3`.  
 And expose them.

I'll then ask @tbowers7 for help to integrate them into the 
  framework he build for this PR.  Bottom line, defer the masters files
  until a (not so) future PR.  Thx!

tbowers7 committed:
Install doc updates

Describe how to pre-cache GitHub-only files for offline PypeIt use.

The script `pypeit_cache_github_data` now also downloads the TelGrid file
specified in the spectrograph's instrument-wide parameters.

	modified:   doc/installing.rst
	modified:   pypeit/scripts/cache_github_data.py

tbowers7 committed:
Update other docs related to this PR

	new file:   doc/api/pypeit.scripts.cache_github_data.rst
	modified:   doc/api/pypeit.scripts.rst
	modified:   doc/include/dependencies_table.rst

tbowers7 committed:
Address PR comments

* Remove `import..as..` statements
* Scrubbed docstring syntax in `pypeit.data.utils`
* Clarified `wvarxiv` output from `pypeit_identify`
* Changed `copy_to_pkgdir` option to `symlink_in_pkgdir` and limited scope

	modified:   doc/wave_calib.rst
	modified:   pypeit/core/gui/identify.py
	modified:   pypeit/data/utils.py
	modified:   pypeit/scripts/cache_github_data.py
	modified:   pypeit/sensfilearchive.py
	modified:   pypeit/tests/test_sensfilearchive.py

tbowers7 commented:
@kbwestfall :
> * Can you remind me of the use case for the `copy_to_pkgdir` option?  Is it possible/useful to instead have a symlink installed in the package directory?  I'm wary of having multiple files on disk, particularly for the large telluric files.

This option is only necessary for a single unit test case to appease an `assert` statement regarding the name of a `sensfunc` file.  I've removed the option from the `reid_arxiv` and `skisim` functions, and have converted the remaining instance to a symlink situation.

> * For the telluric files, can we eventually get all the telluric files in the cloud?  I think I remember that being the goal, but I don't remember the hang-up.

@profxj has uploaded to the cloud the telluric files specifically called out in spectrograph parameter lists.  I'll defer to him the question of putting the remaining files there.

profxj commented:
There is no problem putting all the `telluric` files in the cloud.
  I simply prioritized the ones required for redux

kbwestfall commented:
Sorry to harp on this, @tbowers7 , but the docstring syntax still won't render correctly.  Here's an example for the `get_reid_arxiv_filepath` function:

Current:
```
    Args:
        arxiv_file: str
          The base filename of the ``reid_arxiv`` file to be located

    Returns:
        calibfile: str
          The full path to the ``reid_arxiv`` file
```
Renders as:
![incorrect_rendering](https://user-images.githubusercontent.com/15785898/165366301-394f506b-c798-4a01-b98d-a23d1b0ae841.png)

Corrected Google style:
```
    Args:
        arxiv_file (str):
            The base filename of the ``reid_arxiv`` file to be located

    Returns:
        str: The full path to the ``reid_arxiv`` file
```
Renders as:
![correct_google_rendering](https://user-images.githubusercontent.com/15785898/165366519-70e607c8-8d98-4d31-8546-591b90c59fd5.png)

Corrected Numpy Style:
```
    Parameters
    ----------
    arxiv_file: str
        The base filename of the ``reid_arxiv`` file to be located

    Returns
    -------
    calibfile: str
        The full path to the ``reid_arxiv`` file
```
![correct_numpy_rendering](https://user-images.githubusercontent.com/15785898/165367030-46fbddc4-7b57-423d-a129-386c2b0fad83.png)

Everything else looks great!

tbowers7 committed:
Docstring syntax

	modified:   pypeit/data/utils.py

tbowers7 commented:
@kbwestfall : Thanks for your patience with the docstring syntax.  Sphinx is still on my "to learn" list.

kbwestfall commented:
Thanks for being willing to bear with me...  There's one final issue, which is a subtle difference between the return syntax for Google and numpy styles.  Google style doesn't allow you define the name of the returned object, whereas you can with numpy format; i.e., the fact that I did not include the name of the returned object in the corrected syntax above is important for the former.  You're only allowed to define the type and provide a description in Google style.  This means that when you're returning multiple objects, the return type is always `tuple`, and you provide the components of the tuple in the description.  If you think it's important to provide the object names in the return statement, you have to use Numpy style.

tbowers7 committed:
More Sphinx; CHANGES

	modified:   CHANGES.rst
	modified:   pypeit/data/utils.py

tbowers7 commented:
If it's still not right, maybe you can expose this branch on readthedocs so I can continue tweaking until it's right.

I usually use Numpy style (but not with Sphinx) for my docstrings, but was trying to match existing PypeIt style.

tbowers7 committed:
Fix unit test fail

	modified:   pypeit/core/gui/identify.py
	modified:   pypeit/tests/test_scripts.py

kbwestfall commented:
The quick way to check how things render is to make the html locally:
```
cd doc
make clean ; make html
open _build/html/index.html
```
and then navigate to the relevant page.  All the function docs are accessed through the `Python API` link in the navigation bar.  The `make html` command also runs a bunch of scripts to create some of the auto-documented files (like the datamodel and parameters).  If you don't need to remake these (i.e., if you're just editing function doc strings), you can speed up the process by just calling `make htmlonly`, instead of `make clean ; make html`.  This still takes a while, but it should be faster than waiting for an exposed branch on readthedocs to rebuild.

The "PypeIt style" is a bit all over the place.  I prefer Google style, Joe prefers Numpy style, and there are still many functions with syntax errors.  If you're more comfortable with Numpy style, I would stick with it and just make sure that you use syntax that Sphinx can correctly render in html.

I'll go ahead and approve now, but it looks like there are some CI failures that should be fixed before merging.

Thanks again for all your effort on this!
